### ISSUE TITLE
lint cache vs spurious MD057 errors

### ISSUE THREAD

**Issue by lviollette** [rvben/rumdl#640]
## Context

I'm using rumdl in CI on a project that uses git submodules. The CI pipeline does a sparse checkout: submodule directories are not checked out, so any markdown link pointing into them would legitimately fail MD057. Those links are intentionally suppressed with `<!-- rumdl-disable MD057 -->` blocks.

The pre-commit configuration runs three hooks in sequence: `trailing-whitespace` → `rumdl check` → `rumdl fmt`. In a failing CI run, `rumdl check` passed cleanly (MD057 correctly suppressed), but `rumdl fmt` emitted spurious `::error` MD057 annotations for the same links that `rumdl check` had just accepted — causing GitHub Actions to display errors on the PR.

## Summary

When `rumdl fmt` (or `rumdl check`) gets a **lint cache hit** but the **workspace-index cache is absent or stale**, the function `build_file_index_only` is called to rebuild the cross-file index. This function correctly extracts cross-file links but has two gaps that cause **MD057 to fire despite being suppressed**:

1. It does not populate inline-disable suppression data (`persistent_transitions`, `file_disabled_rules`, `line_disabled_rules`), so `<!-- rumdl-disable MD057 -->` / `<!-- rumdl-enable MD057 -->` blocks are ignored.
2. It receives the **unfiltered** rules list, so `per-file-ignores` entries from `.rumdl.toml` are ignored — MD057 link data is contributed even for files where MD057 is per-file-ignored.

Both issues share the same root cause: the cache-hit fast path in `process_file_inner` skips the filtering and inline-config parsing that the normal (`lint_and_index`) path performs.

## Affected versions

rumdl 0.1.94, 0.1.95 (latest at time of writing)

### Case 1: inline `<!-- rumdl-disable MD057 -->`

```bash
mkdir /tmp/rumdl-bug && cd /tmp/rumdl-bug

# Anchor project root so rumdl stores cache here
touch .rumdl.toml

cat > test.md << 'EOF'
# Test

<!-- rumdl-disable MD057 -->
[broken link](nonexistent.md)
<!-- rumdl-enable MD057 -->

  - badly indented (MD007 violation, makes the file fixable)
EOF

# Step 1: populate lint cache
rumdl check test.md || true

# Step 2: delete workspace-index cache only
rm -f .rumdl_cache/workspace_index.bin

# Step 3: fmt triggers build_file_index_only — inline disable is ignored
rumdl fmt test.md
```

**Expected:** `rumdl fmt` reports and fixes only the MD005/MD007 indentation issue; no MD057 output.

**Actual:** MD057 fires for the link inside the disable block:
```
test.md:7:1: [MD005] Expected indentation of 0 spaces, found 2 [fixed]
test.md:4:15: [MD057] Relative link 'nonexistent.md' does not exist

Fixed: Fixed 1/2 issues in 1 file
```

---

### Case 2: `per-file-ignores` in `.rumdl.toml`

```bash
mkdir /tmp/rumdl-bug2 && cd /tmp/rumdl-bug2

cat > .rumdl.toml << 'EOF'
[per-file-ignores]
"test.md" = ["MD057"]
EOF

cat > test.md << 'EOF'
# Test

[broken link](nonexistent.md)

  - badly indented (MD007 violation, makes the file fixable)
EOF

# Step 1: populate lint cache (MD057 is filtered out by per-file-ignores)
rumdl check test.md || true

# Step 2: delete workspace-index cache only
rm -f .rumdl_cache/workspace_index.bin

# Step 3: fmt triggers build_file_index_only with UNFILTERED rules —
#         MD057 contribute_to_index runs despite per-file-ignores
rumdl fmt test.md
```

**Expected:** `rumdl fmt` reports and fixes only the MD005/MD007 indentation issue; no MD057 output.

**Actual:** MD057 fires despite the `per-file-ignores` entry:
```
test.md:5:1: [MD005] Expected indentation of 0 spaces, found 2 [fixed]
test.md:3:15: [MD057] Relative link 'nonexistent.md' does not exist

Fixed: Fixed 1/2 issues in 1 file
```

## Root cause

In `src/file_processor/processing.rs`, the lint-cache-hit path calls `build_file_index_only` before `per-file-ignores` filtering is applied:

```rust
// Cache hit path (~line 748): unfiltered `rules` — per-file-ignores not yet applied
rumdl_lib::build_file_index_only(
    &content,
    rules,   // ← unfiltered; get_ignored_rules_for_file() runs AFTER this returns
    flavor,
    Some(std::path::PathBuf::from(file_path)),
)

// Normal path (~line 802): filtered rules, inline config parsed correctly
rumdl_lib::lint_and_index(&content, &filtered_rules, verbose, flavor, source_file, Some(config))
```

`build_file_index_only` in `src/lib.rs` also never calls `inline_config.export_for_file_index()`, so `persistent_transitions` (used for range-based disable/enable blocks) stays empty:

```rust
// lint_and_index does this — build_file_index_only does not:
let (file_disabled, persistent_transitions, line_disabled) = inline_config.export_for_file_index();
file_index.file_disabled_rules = file_disabled;
file_index.persistent_transitions = persistent_transitions;
file_index.line_disabled_rules = line_disabled;
```

`run_cross_file_checks` then calls `file_index.is_rule_disabled_at_line(rule_name, line)`, which binary-searches `persistent_transitions`. With the field empty, it always returns `false` — every MD057 warning passes through regardless of inline disable blocks.

## Proposed fix

**Fix 1 — `build_file_index_only` (`src/lib.rs`):** export inline config data, same as `lint_and_index`:

```rust
let lint_ctx = LintContext::new(content, flavor, source_file);

// ADD: populate inline disable data so is_rule_disabled_at_line works
let (file_disabled, persistent_transitions, line_disabled) = lint_ctx.inline_config().export_for_file_index();
file_index.file_disabled_rules = file_disabled;
file_index.persistent_transitions = persistent_transitions;
file_index.line_disabled_rules = line_disabled;
```

**Fix 2 — cache-hit path (`src/file_processor/processing.rs`):** apply `per-file-ignores` filtering before calling `build_file_index_only`:

```rust
let ignored_rules_for_file = config.get_ignored_rules_for_file(Path::new(file_path));
let rules_for_index: Vec<_> = if !ignored_rules_for_file.is_empty() {
    rules.iter()
        .filter(|r| !ignored_rules_for_file.contains(r.name()))
        .map(|r| dyn_clone::clone_box(&**r))
        .collect()
} else {
    rules.to_vec()
};
rumdl_lib::build_file_index_only(&content, &rules_for_index, flavor, source_file)
```

## Verified fix

Both fixes were applied to the 0.1.95 source and built in an isolated Docker container (`rust:1.94.0-slim`). Running the two repros against the patched binary:

**Case 1 — before fix:**
```
test.md:7:1: [MD005] ... [fixed]
test.md:4:15: [MD057] Relative link 'nonexistent.md' does not exist

Fixed: Fixed 1/2 issues in 1 file
```

**Case 1 — after fix:**
```
test.md:7:1: [MD005] ... [fixed]

Fixed: Fixed 1/1 issues in 1 file
```

**Case 2 — before fix:**
```
test.md:5:1: [MD005] ... [fixed]
test.md:3:15: [MD057] Relative link 'nonexistent.md' does not exist

Fixed: Fixed 1/2 issues in 1 file
```

**Case 2 — after fix:**
```
test.md:5:1: [MD005] ... [fixed]

Fixed: Fixed 1/1 issues in 1 file
```

MD057 is correctly suppressed in both cases after the fix.

## Context (CI observation that found the bug)

This was observed in CI running `pre-commit` with hooks in order: `trailing-whitespace` → `rumdl check` → `rumdl fmt`.

1. `trailing-whitespace` modifies a file (hash H0 → H1)
2. `rumdl check` runs on H1: no MD057 errors — submodule links correctly suppressed by `<!-- rumdl-disable MD057 -->` blocks; workspace-index saved
3. `rumdl fmt` runs on H1: lint cache hit → applies MD007 fix → cross-file check fires MD057 for the same inline-disabled links

The minimal repros above reliably reproduce both cases by deleting `workspace_index.bin` after populating the lint cache.

**Workaround:** pass `--no-cache` to rumdl in CI. This forces `lint_and_index` to always run, which always populates inline disable data correctly and uses filtered rules.

---

> **Note:** The source code investigation, root cause analysis, proposed fix, and verification were done with the help of an AI coding assistant (Claude). This was collaborative work — the bug was found and triaged by a human, the deep code digging by the AI.

**Comment by rvben** [comment]
Fixed in **v0.2.0**. Inline `rumdl-disable` blocks and `per-file-ignores` are now honored on the lint-cache fast path, so a cache hit with a missing/stale `workspace_index.bin` no longer resurrects suppressed MD057 warnings during `fmt`. Closing as released.
